org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FormatterHandler.java

```diff
@@ -114,6 +117,26 @@ private List<org.eclipse.lsp4j.TextEdit> format(ICompilationUnit cu, IDocument d
 		String sourceToFormat = document.get();
 		int kind = getFormattingKind(cu, includeComments);
 		TextEdit format = formatter.format(kind, sourceToFormat, region.getOffset(), region.getLength(), 0, lineDelimiter);
+		int length = region.getLength();
+		try {
+			if (document.getChar(region.getOffset() + length) == NEW_LINE) {
+				for (int i = region.getOffset() + length - 1; i > 0; i--) {
+					char lastCharacter = document.getChar(i);
+					if (Character.isWhitespace(lastCharacter)) {
+						continue;
+					} else if (lastCharacter == COMMA) {
+						String newText = SPACE.repeat(options.getTabSize() * 2);

```

Comment:
I think your problem is this line. This represents indentation. However you're forcing the indentation to always be a "space" and repeating it 2 * (tab size). In my case, I'm using spaces, with tabsize of 4, so I get 8 spaces. This is why I'm seeing the line being double indented.

However this value should depend on :
- whether the client is using spaces or tabs
- the tab size (you're already doing this)

The values of these things are all sent over by the client in the parameters.

Reply:
Yes, I was double indenting to simulate the default formatter's behaviour. Also, I believe tabSize is defined by number of spaces (according to https://javadoc.io/doc/org.eclipse.lsp4j/org.eclipse.lsp4j/0.19.1/org/eclipse/lsp4j/FormattingOptions.html), which is why I used a space character. I think the problem I was experiencing lies in the FormattingOptions misrepresenting the tabSize (likely as communicated by the client).